### PR TITLE
Graduating GRPCRoute to GA

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind documentation
/kind feature

**What this PR does / why we need it**:
This PR graduates GRPCRoute to GA. This means the following:

1. GRPCRoute has a new `v1` API version that is identical to `v1alpha2`
2. Storage version is set to `v1` in both standard and experimental channel - note that this makes it impossible to roll back to v1.0 CRDs, but since this is otherwise a no-op change, I think that's acceptable
3. `v1alpha2` is served in experimental channel but is not served in standard channel. This is primarily to ease the migration for anyone upgrading from earlier usage of GRPCRoute while also avoiding any usage of alpha APIs in standard channel.

**Does this PR introduce a user-facing change?**:
```release-note
GRPCRoute has graduated to GA. We recommend using the "v1" API version with this API now. The "v1alpha2" version of this API is deprecated and will be removed in the future.
```

### CODE REVIEW

File: apis/v1/grpcroute_types.go
Comment (robscott) [inline apis/v1/grpcroute_types.go:1]: For the record, I used `git mv` here to try to maintain git history. That appears to have worked everywhere but the GitHub UI.
Screenshot 2024-04-01 at 3 18 01 PM

**Comment by robscott** [comment]
This is a big PR, adding a hold until we've got multiple LGTMs or approvals.

/hold

**Comment by kflynn** [comment]
Looks good here!
/approve

**Review by shaneutt — COMMENTED** [review]
/lgtm

/cc @gnossen

**Comment by keithmattix** [comment]
> v1alpha2 is served in experimental channel but is not served in standard channel. **This is primarily to ease the migration for anyone upgrading from earlier usage of GRPCRoute** while also avoiding any usage of alpha APIs in standard channel.

How does _not_ serving v1alpha2 in standard channel help migrations?

**Comment by kflynn** [comment]
@robscott, ultimately I think we're falling a bit into the how-before-what trap here -- I kind of feel like we're wrangling about how to do things without having a clear sense of exactly what we need to support. Could we back up a moment and lay out some use cases here? 

A few that come immediately to my mind:

1. The experimental channel has GRPCRoute `v1alpha2` which should be promoted to `v1`. Ana has been installing the experimental channel in her clusters, since she's developing against `GRPCRoute`; she would rather use the standard channel instead, though. What should she do?
2. Same situation as #1, but Ian has been managing the CRDs for Ana. How does the migration happen smoothly?
3. Ana has been using the experimental channel's support for the `TeapotResponse` stanza in HTTPRoute `v1alpha7`. `TeapotResponse` is being promoted to standard channel, and Ana wants to stop installing the experimental channel CRDs. How does `TeapotResponse` get moved to standard exactly? What does Ana need to do? (Assume that we currently have HTTPRoute `v1` as standard.)
4. Same situation as #3, but again, Ian is managing the CRDs instead of Ana having control over this.
5. Shortly before `TeapotResponse` was proposed, the Bertrand Gateway controller wrote GEP-BERTRAND proposing `RussellsTeapot` with the semantics that eventually became `TeapotResponse`. GEP-BERTRAND was accepted into the experimental channel, but now that `TeapotResponse` is the accepted standard, the Bertrand folks need to cope with the fact that they have users of `RussellsTeapot` that need to be migrated to `TeapotResponse`. What should they do, exactly? Assume that they're catching this while `TeapotResponse` is still experimental, and that, again, we currently have HTTPRoute `v1` as standard.
6. Same as #5, but suppose that the Bertrand folks wait until `TeapotResponse` has been promoted to standard.

#1 is, I think, what we've been discussing with GRPCRoute.
#3 is, I think, a hypothetical that @robscott proposed, with concrete names and versions so we can talk about concrete solutions.
#5 is a thing that happened to me recently with a Linkerd-specific CRD, and is likely about to happen to Envoy Gateway with TLS validation.

What other situations come to mind?

**Comment by robscott** [comment]
> @robscott, ultimately I think we're falling a bit into the how-before-what trap here -- I kind of feel like we're wrangling about how to do things without having a clear sense of exactly what we need to support. Could we back up a moment and lay out some use cases here?

Agreed, in my opinion, we're trying to accomplish the following goals:

1. Optimize for stability in standard channel. This means avoiding exposure of alpha API versions in standard channel to avoid future painful deprecations or long term support of alpha.
2. Do what we can to enable users that have tried an experimental channel CRD to have an upgrade path to standard channel. As #2912 shows, this is actually rather difficult with our current model, but is mostly viable in the case of this specific GRPCRoute transition because of how stable the GRPCRoute API has been.

> 1. The experimental channel has GRPCRoute `v1alpha2` which should be promoted to `v1`. Ana has been installing the experimental channel in her clusters, since she's developing against `GRPCRoute`; she would rather use the standard channel instead, though. What should she do?

A. Upgrade to experimental v1.1 CRDs that have both v1alpha2 and v1 API versions
B. Upgrade to controller that supports v1
C. Upgrade to standard channel CRDs

> 2. Same situation as (1), but Ian has been managing the CRDs for Ana. How does the migration happen smoothly?

Same as above, just need to make sure the above order of operations happen, doesn't matter how much time passes in between each step though.

> 3. Ana has been using the experimental channel's support for the `TeapotResponse` stanza in HTTPRoute `v1alpha7`. `TeapotResponse` is being promoted to standard channel, and Ana wants to stop installing the experimental channel CRDs. How does `TeapotResponse` get moved to standard exactly? What does Ana need to do? (Assume that we currently have HTTPRoute `v1` as standard.)

This gets to the root of the problem I'm describing in #2912. There's probably no safe path from experimental to standard in this scenario. HTTPRoute in particular usually has several experimental things at once, and not all of them will graduate at the same time. If you were to upgrade from experimental to standard channel, you'd almost certainly end up losing fields and data from experimental channel with unpredictable outcomes.

> 4. Same situation as (3), but again, Ian is managing the CRDs instead of Ana having control over this.

Still awful unless we split CRDs as proposed in #2912. The current state is that there's no safe path from experimental to standard. GRPCRoute is the exception because the entire resource is graduating and it has been unchanged for many releases.

> 5. Shortly before `TeapotResponse` was proposed, the Bertrand Gateway controller wrote GEP-BERTRAND proposing `RussellsTeapot` with the semantics that eventually became `TeapotResponse`. GEP-BERTRAND was accepted into the experimental channel, but now that `TeapotResponse` is the accepted standard, the Bertrand folks need to cope with the fact that they have users of `RussellsTeapot` that need to be migrated to `TeapotResponse`. What should they do, exactly? Assume that they're catching this while `TeapotResponse` is still experimental, and that, again, we currently have HTTPRoute `v1` as standard.

I think unfortunately you're stuck supporting both APIs for a long time, depending on the support guarantees of your implementation.

> 6. Same situation as (5), but suppose that the Bertrand folks wait until `TeapotResponse` has been promoted to standard.

I think the same largely applies.

**Review by shaneutt — APPROVED** [review]

**Comment by robscott** [comment]
@shaneutt, @kflynn, or anyone else, I think we've got enough approvals, can someone add the final LGTM to merge this in?

/hold cancel

**Comment by kflynn** [comment]
/lgtm

🙂

**Pull request merged**
